guide/src/building_and_distribution/multiple_python_versions.md

```diff
@@ -0,0 +1,60 @@
+# Supporting multiple Python versions
+
+PyO3 supports all actively-supported Python 3 and PyPy versions. As much as possible, this is done internally to PyO3 so that your crate's code does not need to adapt to the differences between each version. However, as Python features grow and change between versions, PyO3 cannot a completely identical API for every Python version. This may require you to add conditional compilation to your crate or runtime checks for the Python version.
+
+This section of the guide first introduces the `pyo3-build-config` crate, which you can use as a `build-dependency` to add additional `#[cfg]` flags which allow you to support multiple Python versions at compile-time.
+
+Second, we'll show how to check the Python version at runtime. This can be useful when building for multiple versions with the `abi3` feature, where the Python API compiled against is not always the same as the one in use.
+
+## Conditional compilation for different Python versions
+
+The `pyo3-build-config` exposes multiple [`#[cfg]` flags](https://doc.rust-lang.org/rust-by-example/attribute/cfg.html) which can be used to conditionally compile code for a given Python version. PyO3 itself depends on this crate, so by using it you can be sure that you are configured correctly for the Python version PyO3 is building against.
+
+This allows us to write code like the following
+
+```rust,ignore
+#[cfg(Py_3_7)]
+fn function_only_supported_on_python_3_7_and_up() { }
+
+#[cfg(not(Py_3_8))]
+fn function_only_supported_before_python_3_8() { }
+
+#[cfg(not(Py_LIMITED_API))]
```

Comment:
wait, is it `Py_LIMITED_API` or `abi3`?

Reply:
`#[cfg(Py_LIMITED_API)]` is what we've been using internally for some time now. I agreeit seems slightly strange at first glance. I think the history of it is that it's designed to be similar to the C preprocessor define:

https://github.com/python/cpython/blob/6cc8ac949907b9a1c0f73709c6978b7a43e634e3/Include/object.h#L232-L234

That does raise a good point though. If any of these cfgs are potentially confusing, we should consider alternatives before we release this to users in 0.14.

Some random ideas:

```rust
#[cfg(py_version = 3_7)]
#[cfg(pyo3 = Py_3_7)]
#[cfg(pyo3 = abi3)]
#[cfg(pyo3 = Py_LIMITED_API)]
#[cfg(pyo3 = PyPy)]
```

The extra verbosity is a little annoying for complex patterns, e.g.

```rust
#[cfg(any(not(pyo3 = Py_LIMITED_API), py_version = Py_3_10))]
```

... but it's not too bad.

Definitely I think that scoping these cfgs would make it much easier for users unfamiliar with PyO3 to figure out where they are coming from. Depends if we think that's worth the extra clutter.

Reply:
One thought is that we also have the (very specialist) `#[cfg(py_sys_config="Py_TRACE_REFS")]` attributes. It might be nicer to move these to just be `#[cfg(Py_TRACE_REFS)]` etc, which then fits nicely with `#[cfg(Py_LIMITED_API)]`.